### PR TITLE
refactor: etcd rpc error compat

### PR DESCRIPTION
**Pull request by bsbds**
Closes: #415

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline/src/etcd_error.rs [hunk 1]
```diff
@@ -83,243 +85,267 @@ impl From<EtcdError> for tonic::Status {
         let (code, message) = match err {
             EtcdError::ErrGRPCEmptyKey => (
                 tonic::Code::InvalidArgument,
-                "etcdserver: key is not provided",
+                "etcdserver: key is not provided".to_owned(),
```
[the remark this answers is not in the document]
Comment (bsbds) [reply xline/src/etcd_error.rs]: Yes,  etcd client uses the entire error string to parse the error type.

File: scripts/validation_test.sh [hunk 2]
```diff
@@ -217,7 +217,7 @@ compact_validation() {
     done
     run_with_expect "${ETCDCTL} get --rev=4 key" "key\nvalue3"
     run_with_expect "${ETCDCTL} compact 5" "compacted revision 5"
-    run_with_match "${ETCDCTL} get --rev=4 key" "required revision 4 has been compacted, compacted revision is 5"
+    run_with_match "${ETCDCTL} get --rev=4 key" "mvcc: required revision has been compacted"
```
[the remark this answers is not in the document]
Comment (bsbds) [reply scripts/validation_test.sh]: It exist in the etcd error description, see https://github.com/etcd-io/etcd/blob/main/api/v3rpc/rpctypes/error.go#L32

File: xline/src/request_validation.rs [hunk 3]
```diff
@@ -177,36 +172,106 @@ impl RequestValidator for AuthRoleAddRequest {
...
+    PasswordEmpty,
+    /// Role name is empty
+    #[error("role name is empty")]
+    RoleEmpty,
+    /// Permission not given
+    #[error("permission not given")]
+    PermissionNotGiven,
 }
 
+// As etcd client use grpc error message to parse error types, this is to make the error
```
Comment (Phoenix500526) [inline xline/src/request_validation.rs:220]: we can link the relevant code here to prove our opinion. It should include the error definitions(error.go) and how etcd client uses these errors (client/doc.go)

**Comment by Phoenix500526** [comment]
Please squash these eight commit into one or two commits.

File: xline/src/request_validation.rs (hunk 3 above)
Comment (bsbds) [reply xline/src/request_validation.rs]: done

**Review by Phoenix500526 — APPROVED** [review]

**Review by iGxnon — APPROVED** [review]

**Pull request merged**